parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectCodecFactory.java

```diff
@@ -0,0 +1,531 @@
+
+    public NoopCompressor() {}
+
+    @Override
+    public BytesInput compress(BytesInput bytes) throws IOException {
+      return bytes;
+    }
+
+    @Override
+    public CompressionCodecName getCodecName() {
+      return CompressionCodecName.UNCOMPRESSED;
+    }
+
+    @Override
+    protected void release() {}
+  }
+
+  static class DirectCodecPool {
+    private static final Log LOG = Log.getLog(DirectCodecPool.class);
+
+    public static final DirectCodecPool INSTANCE = new DirectCodecPool();
+
+    @SuppressWarnings("unchecked")
+    private final Map<CompressionCodec, CodecPool> codecs =
+        (Map<CompressionCodec, CodecPool>) (Object) Collections.synchronizedMap(new HashMap<Object, Object>());
+
+    @SuppressWarnings("unchecked")
+    private final Map<Class<?>, GenericObjectPool> directDePools = (Map<Class<?>, GenericObjectPool>) (Object) Collections
+        .synchronizedMap(new HashMap<Object, Object>());
+    private final Map<Class<?>, GenericObjectPool> dePools = (Map<Class<?>, GenericObjectPool>) (Object) Collections
+        .synchronizedMap(new HashMap<Object, Object>());
+    private final Map<Class<?>, GenericObjectPool> cPools = (Map<Class<?>, GenericObjectPool>) (Object) Collections
+        .synchronizedMap(new HashMap<Object, Object>());
+
+    private DirectCodecPool() {}
+
+    public class CodecPool {
+      private final GenericObjectPool compressorPool;
+      private final GenericObjectPool decompressorPool;
+      private final GenericObjectPool directDecompressorPool;
+      private final boolean supportDirectDecompressor;
+
+      private CodecPool(final CompressionCodec codec){
+        try {
+          boolean supportDirectDecompressor = codec.getClass() == DIRECT_DECOMPRESSION_CODEC_CLASS;
+          compressorPool = new GenericObjectPool(new BasePoolableObjectFactory() {
+            public Object makeObject() throws Exception {
+              return codec.createCompressor();
+            }
+          }, Integer.MAX_VALUE);
+
+          Object com = compressorPool.borrowObject();
+          if (com != null) {
+            cPools.put(com.getClass(), compressorPool);
+            compressorPool.returnObject(com);
+          } else {
+            if (Log.WARN) {
+              LOG.warn("Unable to find ByteBuffer compressor for codec " + codec.getClass().getName());
+            }
+          }
+
+          decompressorPool = new GenericObjectPool(new BasePoolableObjectFactory() {
+            public Object makeObject() throws Exception {
+              return codec.createDecompressor();
+            }
+          }, Integer.MAX_VALUE);
+
+          Object decom = decompressorPool.borrowObject();
+          if (decom != null) {
+            dePools.put(decom.getClass(), decompressorPool);
+            decompressorPool.returnObject(decom);
+          } else {
+            if (Log.WARN) {
+              LOG.warn("Unable to find ByteBuffer decompressor for codec " + codec.getClass().getName());
```

Comment:
I don't want this to go unnoticed completely, but I agree we shouldn't expose warnings to users if there isn't a problem. I'll make the message a little clearer (to state that it will be using the old hadoop API and byte arrays) and move these to debug messages.